### PR TITLE
Added the ability to submit to the register arbitrary class Ads

### PR DESCRIPTION
**Pull request by efajardo**
@bbockelm or @edquist could you review this, please? This was done for [SOFTWARE-2714](https://jira.opensciencegrid.org/browse/SOFTWARE-2714)

### CODE REVIEW

File: condor/condor_meter [hunk 1]
```diff
@@ -711,7 +710,19 @@ def classadToJUR(classad):
 
     setIfExists(r.Queue, classad, "JobUniverse", "Condor's JobUniverse field", setstr=True)
     setIfExists(r.NodeCount, classad, 'MaxHosts', "max")
-
+    ########################################################################################
+    ########################################################################################
+    # Code added to send to arbitrary Ads SOFTWARE-2714
+    ArbitraryJobAds = GratiaCore.Config.getConfigAttribute("ArbitraryJobAdAttributes")
+    DebugPrint(5, "ArbitraryJobAd: %s" % str(ArbitraryJobAds))
+    Arbirtray_adlist = ArbitraryJobAds.split(",")
```
Comment (bbockelm) [inline condor/condor_meter]: This will not handle space-separated lists (`foo bar baz` or `foo, bar, baz`) as is traditional in HTCondor.

Try:
```
import re
...
Arbitrary_adlist = re.split(",?\s?", ArbitraryJobAds)
```

File: condor/condor_meter [hunk 2]
```diff
@@ -711,7 +710,19 @@ def classadToJUR(classad):
...
+    DebugPrint(5, "Going through list: %s " % Arbirtray_adlist)
+    for arbitaryAd in  Arbirtray_adlist:
```
Comment (bbockelm) [inline condor/condor_meter]: Variable name `Arbirtray_adlist` is misspelled.  There's also a mix of CamelCase and underscores - looking at the nearby code, stick with CamelCase.

File: condor/condor_meter [hunk 3]
```diff
@@ -711,7 +710,19 @@ def classadToJUR(classad):
...
+        if arbitaryAd in classad:
```
Comment (bbockelm) [inline condor/condor_meter]: Might consider renaming this to `arbitraryAttr` instead of `arbitraryAd` since it isn't actually an ad...

File: condor/condor_meter [hunk 4]
```diff
@@ -711,7 +710,19 @@ def classadToJUR(classad):
...
+            DebugPrint(5, "Arbitrary ad: %s found in classAd" % arbitaryAd) 
+            DebugPrint(5, "Value of arbitrary classAd: %s" % str(classad[str(arbitaryAd)]))
+            r.AdditionalInfo(arbitaryAd, str(classad[str(arbitaryAd)]))
```
Comment (bbockelm) [inline condor/condor_meter]: For complex expressions, this will simply convert the expression to a string instead of evaluate the expression.

I think we want to evaluate the expression.

File: condor/condor_meter [hunk 5]
```diff
@@ -711,7 +710,19 @@ def classadToJUR(classad):
...
+    ########################################################################################
+    # Code added to send to arbitrary Ads SOFTWARE-2714
+    ArbitraryJobAds = GratiaCore.Config.getConfigAttribute("ArbitraryJobAdAttributes")
+    DebugPrint(5, "ArbitraryJobAd: %s" % str(ArbitraryJobAds))
+    Arbirtray_adlist = ArbitraryJobAds.split(",")
+    DebugPrint(5, "Going through list: %s " % Arbirtray_adlist)
+    for arbitaryAd in  Arbirtray_adlist:
+        if arbitaryAd in classad:
+            DebugPrint(5, "Arbitrary ad: %s found in classAd" % arbitaryAd) 
+            DebugPrint(5, "Value of arbitrary classAd: %s" % str(classad[str(arbitaryAd)]))
```
Comment (bbockelm) [inline condor/condor_meter]: Convert the two lines of debug statements into a single one.

File: condor/condor_meter [hunk 6]
```diff
@@ -711,7 +710,19 @@ def classadToJUR(classad):
 
     setIfExists(r.Queue, classad, "JobUniverse", "Condor's JobUniverse field", setstr=True)
     setIfExists(r.NodeCount, classad, 'MaxHosts', "max")
-
+    ########################################################################################
+    ########################################################################################
+    # Code added to send to arbitrary Ads SOFTWARE-2714
+    ArbitraryJobAds = GratiaCore.Config.getConfigAttribute("ArbitraryJobAdAttributes")
```
Comment (bbockelm) [inline condor/condor_meter]: The config variable name is a mouthful.  How about `ExtraAttributes`?

File: condor/condor_meter (hunk 1 above)
Comment (osg-cat) [inline condor/condor_meter]: Won’t that allow it to split every single character apart? Perhaps something like this:

`Arbitrary_adlist = re.split(r'[,\s]+', ArbitraryJobAds)`

Comment (bbockelm) [inline condor/condor_meter]: Those are equivalent.

Comment (osg-cat) [inline condor/condor_meter]: They are not equivalent regular expressions, but I guess that they behave similarly in this particular case due to a oddity (IMO) of Python. The docs for re.split() say, “Note that split will never split a string on an empty pattern match”, which is key here.

**Comment by efajardo** [comment]
@bbockelm,

I did the changes you requested.

For testing purposes I modify this file:

`/etc/gratia/condor/ProbeConfig`

With the following:

```Title2="Probe information and functional configuration"
ProbeName="condor:uclhctest.t2.ucsd.edu"
SiteName="UCSD"
Grid="OSG"
EnableProbe="1"
...
ExtraAttributes="foo,foo2,foo3"
```

I then submitted a job with the following submit file

```
output = myoutput.$(Cluster)
error = myerror.$(Cluster)
log = mylog.$(Cluster)
executable = echoScript.sh
+foo = "bar"
+foo2 = "bar2"
+foo3 = "bar3"
queue 10
```

And checked here in [GRAC](https://gracc.opensciencegrid.org/kibana/app/kibana?#/discover?_g=(refreshInterval:(%27$$hashKey%27:%27object:267%27,display:%275%20seconds%27,pause:!t,section:1,value:5000),time:(from:now%2Fw,mode:quick,to:now%2Fw))&_a=(columns:!(ProbeName),index:%27gracc.osg.raw-*%27,interval:auto,query:(query_string:(analyze_wildcard:!f,query:%27ProbeName:%22condor:uclhctest.t2.ucsd.edu%22%27)),sort:!(EndTime,desc))), that the records have:

```? Resource_foo	   	  bar
? Resource_foo2	   	  bar2
? Resource_foo3	   	  bar3
```

Is this allright?

File: condor/condor_meter [hunk 7]
```diff
@@ -711,7 +710,18 @@ def classadToJUR(classad):
...
     setIfExists(r.Queue, classad, "JobUniverse", "Condor's JobUniverse field", setstr=True)
     setIfExists(r.NodeCount, classad, 'MaxHosts', "max")
-
+    ########################################################################################
+    ########################################################################################
+    # Code added to send to arbitrary Ads SOFTWARE-2714
+    ArbitraryJobAttrs = GratiaCore.Config.getConfigAttribute("ExtraAttributes")
+    DebugPrint(5, "Arbitrary Job Attributes: %s" % str(ArbitraryJobAttrs))
+    ArbitraryAttrslist = re.split(",?\s?", ArbitraryJobAttrs)
+    #ArbitraryAttrslist = ArbitraryJobAds.split(",")
```
Comment (bbockelm) [inline condor/condor_meter]: Please remove commented-out line.

File: condor/condor_meter [hunk 8]
```diff
@@ -770,6 +780,7 @@ def classadToJUR(classad):
         if isinstance(evaluated, types.StringType):
             r.SiteName(evaluated)
             r.ProbeName('%s-%s' % (r.GetProbeName(), evaluated))
+
```
Comment (bbockelm) [inline condor/condor_meter]: Please remove unnecessary whitespace change.
